src/charm.py

```diff
@@ -9,93 +9,306 @@
+        if not self.topic_name and not self.database_name:
+            return BlockedStatus("Please specify either topic or database name")
 
-        self._on_config_changed(None)
+        if not self.is_database_related and not self.is_kafka_related:
+            return BlockedStatus("Please relate the data-integrator with the desired product")
+
+        if self.is_kafka_related and self.topic_active != self.topic_name:
+            logger.error("Trying to change Kafka configuration for existing relation")
+            return BlockedStatus("To change topic, please remove relation and add it again")
+
+        if self.is_database_related and any(
+            [database != self.database_name for database in self.databases_active.values()]
+        ):
+            logger.error("Trying to change database-name configuration for existing relation")
+            return BlockedStatus(
+                "To change database name, please remove relation and add it again"
+            )
+
+        return ActiveStatus()
+
+    def _on_config_changed(self, _) -> None:
+        """Handle on config changed event."""
+        # Only execute in the unit leader
+        if not self.unit.is_leader():
+            return
 
-    def _on_config_changed(self, event):
-        self.database = self.model.config.get("database", "")
-        if self.database == "":
-            self.unit.status = BlockedStatus("The database name is not specified.")
-        elif event:
-            self.unit.status = ActiveStatus(f"database: {self.database}")
+        self.unit.status = self.get_status()
 
-        self.mysql.database = self.database
-        self.postgresql.database = self.database
-        self.mongodb.database = self.database
+        # update relation databag
+        # if a relation has been created before configuring the topic or database name
+        # update the relation databag with the proper value
+        if not self.databases_active and self.database_name:
+            database_relation_data = {
+                "database": self.database_name,
+                "extra-user-roles": self.extra_user_roles or "",
+            }
+            self._update_database_relations(database_relation_data)
 
-        for rel in self.mysql.relations:
-            self.mysql._update_relation_data(rel.id, {"database": self.database})
-        for rel in self.postgresql.relations:
-            self.postgresql._update_relation_data(rel.id, {"database": self.database})
-        for rel in self.mongodb.relations:
-            self.mongodb._update_relation_data(rel.id, {"database": self.database})
+        if not self.topic_active and self.topic_name:
+            for rel in self.kafka.relations:
+                self.kafka._update_relation_data(
+                    rel.id,
+                    {
+                        "topic": self.topic_name or "",
+                        "extra-user-roles": self.extra_user_roles or "",
+                    },
```

Comment:
**thought:** Non-blocking, just an idea, I wonder if this could take a `DefaultDict`, do some clever star-expansion from `**self`, and any missing/None attrs defaults dict values to to `""`. Complicated and not needed though.

Reply:
I'd defer this later. This PR is already quite big to also put extra features